ygen/enumgen.go

```diff
@@ -61,29 +60,66 @@ type enumGenState struct {
 	// a module such as openconfig-bgp which defines /bgp and is also used at
 	// /network-instances/network-instance/protocols/protocol/bgp.
 	uniqueEnumeratedLeafNames map[string]string
+
+	// The flags below are cached findEnumSet parameters that are required for
+	// re-calculating the keys used to query the name sets.
+	compressPaths bool
+	noUnderscores bool
+	skipDedup     bool
 }
 
-// newEnumGenState creates a new enumGenState instance initialised with the
-// default state required for code generation.
-func newEnumGenState() *enumGenState {
-	return &enumGenState{
-		definedEnums:                 map[string]bool{},
+// newEnumSet initializes a new empty enumSet instance.
+func newEnumSet(compressPaths, noUnderscores, skipEnumDedup bool) *enumSet {
+	return &enumSet{
 		uniqueIdentityNames:          map[string]string{},
 		uniqueEnumeratedTypedefNames: map[string]string{},
 		uniqueEnumeratedLeafNames:    map[string]string{},
+		compressPaths:                compressPaths,
+		noUnderscores:                noUnderscores,
+		skipDedup:                    skipEnumDedup,
+	}
+}
+
+// identityRefBaseType retrieves the mapped name of an identityref's
+// base such that it can be used in generated code. The value that is returned
+// is defining module name followed by the CamelCase-ified version of the
+// base's name. This function wraps the identityrefBaseTypeFromIdentity
+// function since it covers the common case that the caller is interested in
+// determining the name from an identityref leaf, rather than directly from the
+// identity.
+func (s *enumSet) identityRefBaseType(idr *yang.Entry) (string, error) {
```

Comment:
Should we consider calling this `generatedNameOfBaseFromLeaf` or something? I think that `identityRefBaseType` is what I'd expect the `...FromIdentity` case of this to call.